### PR TITLE
Require timestamps for importmulti keys

### LINKED ISSUE #9491: Importmulti api is confusing in a way that could lead to funds loss.
Our normal import interfaces introduce keys with an effective timestamp of 0.

Importmulti introduces keys with a default effective timestamp of 'now'.

While documented it is highly surprising, especially because it contradicts the default on the rescan argument:

>     "rescan": <false>,         (boolean, optional, default: true) Stating if should rescan the blockchain after all imports

This means you can import a key and reasonably expect that it was rescanned ("I even set rescan true!") but have it not be rescanned, causing you to think the key had been unused and then you give it away and or destroy it and lose funds.

There are a number of other bugs in importmulti (mishandles pruning, needlessly forces the wallet to be unlocked when not importing private keys, etc.) which I am fixing, but I need feedback on this API issue.

Should the default become zero, with -1 special cased as 'now' or make timestamp mandatory? default it to the prune depth??

Should the result also tell us what it rescanned?

### PR DESCRIPTION
**Pull request by ryanofsky**
Require timestamps for importmulti keys

Additionally, accept a "now" timestamp, to allow avoiding rescans for keys
which are known never to have been used.

Note that the behavior when "now" is specified is slightly different than the
previous behavior when no timestamp was specified at all. Previously, when no
timestamp was specified, it would avoid rescanning during the importmulti call,
but set the key's nCreateTime value to 1, which would not prevent future block
reads in later ScanForWalletTransactions calls. With this change, passing a
"now" timestamp will set the key's nCreateTime to the current block time
instead of 1.

Fixes #9491

### CODE REVIEW

**Comment by TheBlueMatt** [comment]
Needs 0.14 tag.

File: src/wallet/rpcdump.cpp [hunk 1]
```diff
@@ -970,13 +984,16 @@ UniValue importmulti(const JSONRPCRequest& mainRequest)
             "  [     (array of json objects)\n"
             "    {\n"
             "      \"scriptPubKey\": \"<script>\" | { \"address\":\"<address>\" }, (string / json, required) Type of scriptPubKey (string for script, json for address)\n"
+            "      \"timestamp\": 1454686740 | \"now\"                       , (integer / string) Integer timestamp, or the string \"now\" to substitute the current synced\n"
```
Comment (achow101) [inline src/wallet/rpcdump.cpp]: You should say whether it is required or optional

**Comment by gmaxwell** [comment]
Concept ACK but people with more opinions (and knowledge about json rpc norms) should comment on the type of the timestamp argument being variable.

File: src/wallet/rpcdump.cpp (hunk 1 above)
Comment (jonasschnelli) [inline src/wallet/rpcdump.cpp]: If we are at this, why do we have the number `1454686740` here? Can't we just say `<timestamp>`?

**Comment by jonasschnelli** [comment]
Concept ACK, will test.

I liked the idea from @sipa where he mentioned that "unknown" as timestamp for avoiding rescans seems to be the better choice (instead of "now"). "Now" would imply that the key was generated during import which is (almost) impossible.

**Comment by MarcoFalke** [comment]
sipa proposed `timestamp: "unused"` instead of "now".
https://github.com/bitcoin/bitcoin/issues/9491#issuecomment-277054374

**Comment by laanwj** [comment]
> Concept ACK but people with more opinions (and knowledge about json rpc norms) should comment on the type of the timestamp argument being variable.

It's hard to wrap in bindings for statically typed languages - could be worked around with an extra flag that specifies how to handle the timestamp.

File: src/wallet/rpcdump.cpp (hunk 1 above)
Comment (ryanofsky) [inline src/wallet/rpcdump.cpp]: Updated in af06509f9c0bb92231ab41fd5fd9e89106b65b0e. Personally I think `"timestamp": 1454686740` is more informative than `"timestamp": <timestamp>`, made the requested change anyway.

**Comment by ryanofsky** [comment]
Added RPC test to verify that this is correctly interpreting the "now" value and saving the current time as the key creation time in 2cd875a53ed4ea7bde36f943603949f30bdcbc30.

**Comment by ryanofsky** [comment]
Squashed 2cd875a53ed4ea7bde36f943603949f30bdcbc30 -> c07ebb6efced42cb9e4ed8bce387288ae705dca8 ([multinow.3](https://github.com/ryanofsky/bitcoin/commits/pr/multinow.3) -> [multinow.4](https://github.com/ryanofsky/bitcoin/commits/pr/multinow.4))

File: src/wallet/rpcdump.cpp [hunk 2]
```diff
@@ -1015,6 +1032,12 @@ UniValue importmulti(const JSONRPCRequest& mainRequest)
     LOCK2(cs_main, pwalletMain->cs_wallet);
     EnsureWalletIsUnlocked();
 
+    // Verify all timestamps are present before importing any keys.
+    const int64_t now = chainActive.Tip() ? chainActive.Tip()->GetBlockTime() : 0;
```
Comment (TheBlueMatt) [inline src/wallet/rpcdump.cpp]: Can we use the MTP here?

**Comment by TheBlueMatt** [comment]
utACK c07ebb6efced42cb9e4ed8bce387288ae705dca8 if we switch to using MTP for the "now" timestamp.

File: src/wallet/rpcdump.cpp (hunk 2 above)
Comment (ryanofsky) [inline src/wallet/rpcdump.cpp]: Switched to MTP in dc402a4444d09263b74483a7d8ed904325dd89f6

**Review by ryanofsky — COMMENTED** [review]
Now using MTP in dc402a4444d09263b74483a7d8ed904325dd89f6

**Comment by TheBlueMatt** [comment]
utACK dc402a4444d09263b74483a7d8ed904325dd89f6

**Comment by laanwj** [comment]
> I liked the idea from @sipa where he mentioned that "unknown" as timestamp for avoiding rescans seems to be the better choice (instead of "now"). "Now" would imply that the key was generated during import which is (almost) impossible.
> sipa proposed `timestamp: "unused"` instead of "now".

A "unused" flag makes sense, however that is not a property of the timestamp.

**Comment by kallewoof** [comment]
`Timestamp = unused` sounds strange to me as well. `Timestamp = now` makes more sense.

File: src/rpc/misc.cpp [hunk 3]
```diff
@@ -167,6 +167,7 @@ UniValue validateaddress(const JSONRPCRequest& request)
             "  \"pubkey\" : \"publickeyhex\",    (string) The hex value of the raw public key\n"
             "  \"iscompressed\" : true|false,  (boolean) If the address is compressed\n"
             "  \"account\" : \"account\"         (string) DEPRECATED. The account associated with the address, \"\" is the default account\n"
+            "  \"timestamp\" : <timestamp>,      (number, optional) The creation time of the key if available\n"
```
Comment (kallewoof) [inline src/rpc/misc.cpp]: Nit: Non-string arguments tend to be represented as simply `name`, not `<name>`, e.g. https://github.com/bitcoin/bitcoin/blame/master/src/rpc/blockchain.cpp#L649

File: src/wallet/rpcdump.cpp [hunk 4]
```diff
@@ -640,7 +640,8 @@ UniValue dumpwallet(const JSONRPCRequest& request)
 }
 
 
-UniValue processImport(const UniValue& data) {
+UniValue processImport(const UniValue& data, int64_t timestamp)
```
Comment (kallewoof) [inline src/wallet/rpcdump.cpp]: Nit: `const int64_t timestamp`
Nit 2 / coding standard: Functions in Bitcoin generally start with a capital letter, i.e. `ProcessImport` here. (Exception being the RPC commands, which match the CLI command names.) I know you didn't name this one, but it coincides with the new function below (`getImportTimestamp`). Renaming this one should not be a big issue, as it is used in one place only.

File: src/wallet/rpcdump.cpp [hunk 5]
```diff
@@ -958,6 +958,20 @@ UniValue processImport(const UniValue& data) {
     }
 }
 
+int64_t getImportTimestamp(const UniValue& data, int64_t now)
```
Comment (kallewoof) [inline src/wallet/rpcdump.cpp]: Nit / coding standard / repeat: `GetImportTimestamp` (see above)

File: src/wallet/rpcdump.cpp [hunk 6]
```diff
@@ -970,13 +984,16 @@ UniValue importmulti(const JSONRPCRequest& mainRequest)
             "  [     (array of json objects)\n"
             "    {\n"
             "      \"scriptPubKey\": \"<script>\" | { \"address\":\"<address>\" }, (string / json, required) Type of scriptPubKey (string for script, json for address)\n"
+            "      \"timestamp\": <timestamp> | \"now\"                      , (integer / string, required) Integer timestamp, or the string \"now\" to substitute the current synced\n"
```
Comment (kallewoof) [inline src/wallet/rpcdump.cpp]: Nit / repeat: `<timestamp>` -> `timestamp` (see above comment)

File: src/wallet/rpcdump.cpp [hunk 7]
```diff
@@ -1015,6 +1032,12 @@ UniValue importmulti(const JSONRPCRequest& mainRequest)
     LOCK2(cs_main, pwalletMain->cs_wallet);
     EnsureWalletIsUnlocked();
 
+    // Verify all timestamps are present before importing any keys.
+    const int64_t now = chainActive.Tip() ? chainActive.Tip()->GetMedianTimePast() : 0;
```
Comment (kallewoof) [inline src/wallet/rpcdump.cpp]: I don't know if 0 is a good default for `now`, but it doesn't seem harmful either. I wonder if the function shouldn't simply throw a `runtime_exception` if `!chainActive.Tip()` saying to sync up first..

File: src/wallet/rpcdump.cpp [hunk 8]
```diff
@@ -970,13 +984,16 @@ UniValue importmulti(const JSONRPCRequest& mainRequest)
             "  [     (array of json objects)\n"
             "    {\n"
             "      \"scriptPubKey\": \"<script>\" | { \"address\":\"<address>\" }, (string / json, required) Type of scriptPubKey (string for script, json for address)\n"
+            "      \"timestamp\": <timestamp> | \"now\"                      , (integer / string, required) Integer timestamp, or the string \"now\" to substitute the current synced\n"
+            "                                                              blockchain time. The timestamp of the oldest key will determine how far back blockchain rescans\n"
+            "                                                              for missing wallet transactions need to begin. \"now\" can be specified to bypass scanning, for\n"
+            "                                                              keys which are known to never have been used, and 0 can be specified to scan the entire blockchain.\n"
```
Comment (kallewoof) [inline src/wallet/rpcdump.cpp]: Language: "The timestamp of the oldest key will determine how far back blockchain rescans need to begin for finding missing wallet transactions." maybe?

**Review by kallewoof — APPROVED** [review]
utACK, did not review tests

**Comment by ryanofsky** [comment]
Squashed 6524d401485a8dbd9881e6d69b3f542bb6598d6b -> 715f05d9987b6b9d8c2737d40a7113dd0680cfc4 ([multinow.6](https://github.com/ryanofsky/bitcoin/commits/pr/multinow.6) -> [multinow.7](https://github.com/ryanofsky/bitcoin/commits/pr/multinow.7))

File: src/wallet/rpcdump.cpp (hunk 7 above)
Comment (kallewoof) [inline src/wallet/rpcdump.cpp]: OK, yeah I think a notification would be useful.

**Review by jonasschnelli — APPROVED** [review]
Tested ACK 715f05d9987b6b9d8c2737d40a7113dd0680cfc4

**Comment by ryanofsky** [comment]
Rebased 715f05d9987b6b9d8c2737d40a7113dd0680cfc4 -> 266a8114cbe2a87a6c84d7690a7716a18d782c56 ([multinow.7](https://github.com/ryanofsky/bitcoin/commits/pr/multinow.7) -> [multinow.8](https://github.com/ryanofsky/bitcoin/commits/pr/multinow.8)) following the #9227 merge.

This PR didn't conflict with #9227, but I'm about to rebase #9108 on top of this, and #9108 did conflict with #9227.

**Comment by morcos** [comment]
utACK 266a811

**Pull request merged**